pam-reg-cc/pam_reg_cc.8
.TH PAM_REG_CC 8
.SH NAME
pam_reg_cc \- register credential caches for renewd
.SH DESCRIPTION
.I  pam_reg_cc [usecollection] [usedefaultname]
.PP
Pam_reg_cc registers Kerberos credentials so that renewd will
renew them. Depending upon the configuration in /etc/krb5.conf,
it may also normalize the name to work around issues with various
versions of sshd.
.PP
Pam_reg_cc checks the current value of KRB5CCNAME in the environment that
PAM is preparing for the user process. If it is defined, 
pam_reg_cc registers the cache that was just created for
renewd to renew automatically.
.PP
pam_reg_cc implements both auth and session. Logically it should be
used in the auth stack after whatever module does Kerberos authentication.
However sshd sets up Kerberos credentially internally, and doesn't
call PAM's auth code. So the normal configuration is to put
it in session fairly near the beginning, before anything that
tries to access the user's home directory (particular on Ubuntu).
Auth will be used primarily when the
application doesn't call session, or when it doesn't have KRB5CCNAME
set for the session.
.PP
In case the application doesn't make KRB5CCNAME available, you can
set "usedefaultname." This will cause pam_reg_cc to assume that the
credential cache uses the default name as defined in /etc/krb5.conf.
This can be dangerous, since sshd and other applications may make their
own choice about the credential cache. But some applications need this.
.PP
KRB5CCNAME is sometimes set
to the default collection, rather than to the specific cache.
That can require pam_reg_cc to guess which cache was actually
just created. It will pick a cache from the collection that
has the user's main principal. That's the same cache that sssd
uses, so it should be the right one.
.PP
If usecollection is specified, and the cache is in a collection
type cache (e.g. KEYRING or KCM), strips
any subsidiary name. E.g if it is KEYRING:persistent:1003:1003,
although that name will be registered for renewd, the environment
variable will be updated to KEYRING:persistent:1003. This will 
allow kinit, klist, and kdestroy to work as intended with collections.
.PP
In general we recommend setting default_ccache_name in krb5.conf to
KEYRING:persistent:%{uid} or KCM:, and setting usecollection. 
.PP
Sshd has two known problems: (1) On Centos it tends to set KRB5CCNAME
to the actual cache name. If the user does kinit as a different
principal, it will overwrite the cache. This can cause the user to
lose access to NFS. To fix this, set the default cc in /etc/krb5.conf
to KEYRING:persistent:%{uid} or KCM:, and set "usecollection." As explained
above, that will fix up KRB5CCNAME to point to the collection. Kinit
is smart enough that it will create a new cache if the user kinits
as a different principal. If used with the gssd wrapper, gssd will
pick the right credential in the cache.
.PP
Sshd (2) On Ubuntu, it uses /tmp/krb5...., even if krb5.conf
sets the default credential cache to something else.
Because /tmp/krb5... isn't a collection, That causes the same
issue as above, but in addition, rpc.gssd doesn't seem to work as
well with credentials in /tmp as in KEYRING. So there's a special
check in pam_reg_cc. If the default in krb5.conf is a collection,
and KRB5CCNAME is something else, it copies the credentials into a new
collection that matches what's in krb5.conf and fixes up KRB5CCNAME to point to it. So on
Ubuntu we still recommend setting the default in krb5.conf to
KEYRING:persistent:%{uid} or KCM:, and setting "usecollection".
.PP
From auth, this module implements a no-op authenticate. The code is in setcred.
From session, it implements session start. Session end is a no-op. It does not 
make sense to use it for both auth and session, though it's normally harmless
.PP
If this module is being used in a system with renewd, you may want to warn users if the
credentials forwarded to them have too short a lifetime to be renewed. To warn if there
is less than 55 min left use. You may also want to specify ticket_warn_file, which is
a file with the text of the warning message to be printed. There's default text in case
it isn't specified, but this lets you specify your own.
.PP
.nf
[appdefaults]
register-cc = {
    ticket_minlife = 55m
    ticket_warn_file = /etc/krb5.ticketwarning
}
.fi
.PP
NOTE: There used to be additional functionality to make a second copy of the
credential cache. This turned out not to be useful, and has been removed.


krb5-strength-3.1/NEWS
                    User-Visible krb5-strength Changes

krb5-strength 3.1 (2016-12-25)

    A new configuration option, cracklib_maxlen, can be set to skip
    CrackLib checks of passwords longer than that length.  The CrackLib
    rules were designed in a world in which most passwords were four to
    eight characters long and tend to spuriously reject longer passwords.
    SQLite dictionaries work better for checking longer passwords and
    passphrases.  Patch from Jorj Bauer.

    The require_classes configuration option can now require a particular
    number of character classes in the password (whatever those classes
    are).  Patch from Toby Blake.

    Change the error messages returned for passwords that fail strength
    checking to start with a capital letter.  This appears to be more
    consistent with the error message conventions used inside Heimdal.

    Change the DB_File::Lock calling method in heimdal-history to work
    properly with the (buggy) CPAN version of DB_File::Lock, instead of
    relying on Debian's patched version.  Thanks to Bernt Jernberg for the
    report.

    Apply the SuSE patch for a buffer overflow when using duplicate rules
    to the embedded CrackLib.  No duplicating rules are used in the rule
    set included with this package, and this package doesn't expose the
    general API, so this was not exploitable, but best to close the latent
    issue.  (The other recent CrackLib vulnerability, CVE-2016-6318,
    doesn't apply since all the GECOS manipulation code was removed from
    the embedded CrackLib in this package.)

    Patch the mkdict and packer in the embedded copy of CrackLib to force
    C locale when sorting (avoiding a corrupted dictionary) and warn and
    skip out-of-order words rather than creating a corrupted dictionary.
    Patch from Mark Sirota.

    Configuration instrutions are now in the heimdal-history and
    heimdal-strength man pages and a new krb5-strength man page (which
    documents configuration of the KDC plugin) instead of the README file
    to make it more accessible after the software has been installed.

    Update to rra-c-util 6.2:

    * Use calloc in preference to malloc wherever appropriate.
    * Use reallocarray in preference to realloc wherever appropriate.
    * Suppress warnings from Kerberos headers under make warnings.
    * Support the embedded Kerberos in Solaris 10 in library probes.
    * Add missing va_end in xasprintf implementation.
    * Fix logic in Test::RRA::Automake for new Automake dist checking.
    * Fix all return-value checks for snprintf to avoid off-by-one error.
    * Update warning flags for make warnings to GCC 6.1.0.
    * Fix Test::RRA::Config for new "do" semantics in Perl 5.22.2.
    * Add a new test for obsolete eyrie.org URLs.
    * Require Test::Strict 0.25 or newer for Perl strictness checks.

    Update to C TAP Harness 4.1:

    * Replace all remaining uses of sprintf.
    * Test lists may now have comments and blank lines.
    * runtests -v will show the complete output from a test.
    * Fix segfault in runtests when given an empty test list.
    * Tests use C_TAP_SOURCE and C_TAP_BUILD instead of SOURCE and BUILD.

krb5-strength 3.0 (2014-03-25)

    The krb5-strength plugin and heimdal-strength program now support a
    SQLite password dictionary.  This format of dictionary can detect any
    password within edit distance one of a dictionary word, meaning that
    the dictionary word can be formed by adding, removing, or changing a
    single character in the password.  A SQLite password dictionary can be
    used alone or in combination with any of the other supported
    dictionary types.  SQLite dictionary support is based on work by David
    Mazières.

    cdbmake-wordlist has been renamed to krb5-strength-wordlist.
    Generating CDB dictionaries now requires the -c option; see the
    documentation for more information.  A SQLite database of dictionary
    words can now be created instead, using the -s option.

    A password history implementation for Heimdal is now included.  This
    is a separate Perl program, heimdal-history, that stacks with the
    external program implementation of strength checking.  It is not
    available in the form of a plugin, only as a Heimdal external password
    quality check.  (MIT Kerberos provides its own password history
    mechanism.)  This program has more extensive Perl module dependencies
    than the other programs in this distribution.

    A new configuration option, minimum_different, can be set to require
    that passwords contain at least that many unique characters.  This can
    be used to reject long strings of identical characters or short
    patterns, which may pass other checks but still be too easy to guess.

    Update to rra-c-util 5.4:

    * Fix portable/krb5.h build with a C++ compiler.
    * Use Lancaster Consensus environment variables to control tests.
    * Work around perltidy bug that leaves behind stray log files.

    Update to C TAP Harness 3.0:

    * Reopen standard input to /dev/null when running a test list.
    * Don't leak extraneous file descriptors to tests.

krb5-strength 2.2 (2013-12-16)

    More complex character class requirements can be specified with the
    configuration option require_classes.  This option lists the character
    classes the password must contain.  These restrictions may be
    qualified with password length ranges, allowing the requirements to
    change with the length of the password.  See README for more details
    and the option syntax.


pam-reg-cc/pam_reg_cc.c
#define PAM_SM_SESSION
#define PAM_SM_AUTH
#define _GNU_SOURCE 

#include <sys/types.h>
#include <sys/errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <security/pam_appl.h>
#include <security/pam_modules.h>
#include <security/pam_ext.h>
#include <syslog.h>
#include "krb5.h"
#include "com_err.h"
#include <pwd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>
#include "../common/ccacheutil.h"

/*
 * Copyright 2017 by Rutgers, the State University of New Jersey
 * All Rights Reserved.
 *
 * Export of this software from the United States of America may
 *   require a specific license from the United States Government.
 *   It is the responsibility of any person or organization contemplating
 *   export to obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of Rutgers not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original Rutgers software.
 * Rutgers makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */


/*

Post process credential cache

We know of three cases:

sshd: uses the primary, even if it's wrong principal
  sets KRB5CCMAME to cache
Ubuntu sshd: always uses /tmp/...
  copy to KEYRING if that's the default, and change KRB5CCNAME
sssd: uses a cache with matching principal, doesn't change primary
  set KRB5CCNAME to collection
cache in /tmp: sets KRB5CCNAME to the file if not cron

*/

static inline int
data_eq(krb5_data d1, krb5_data d2)
{
  return (d1.length == d2.length && (d1.length == 0 ||
				     !memcmp(d1.data, d2.data, d1.length)));
}

static inline int
data_eq_string (krb5_data d, const char *s)
{
  return (d.length == strlen(s) && (d.length == 0 ||
				    !memcmp(d.data, s, d.length)));
}

krb5_boolean is_local_tgt (krb5_principal princ, krb5_data *realm);

/* Return true if princ is the local krbtgt principal for local_realm. */
krb5_boolean
is_local_tgt(krb5_principal princ, krb5_data *realm)
{
  return princ->length == 2 && data_eq(princ->realm, *realm) &&
    data_eq_string(princ->data[0], KRB5_TGS_NAME) &&
    data_eq(princ->data[1], *realm);
}

#define RENEWDCCS "/run/renewdccs/"

void register_for_delete(pam_handle_t *pamh, const char *cache, uid_t uid) {
  char *newname;
  const char *cp;
  char *cp2;
  int fd;

  newname = malloc(strlen(cache) + strlen(RENEWDCCS) + 1);

  strcpy(newname, RENEWDCCS);
  cp2 = newname + strlen(RENEWDCCS);

  // none of our cache names use \, so map / to \

  cp = cache;

  while (*cp) {
    char ch = *cp;
    if (ch == '/')
      *cp2 = '\\';
    else
      *cp2 = ch;
    cp++;
    cp2++;
  }
  *cp2 = '\0';
  fd = open(newname, O_CREAT|O_WRONLY, 0600);
  if (fd < 0 && errno == ENOENT) {
    fd = mkdir(RENEWDCCS, 0700);
    if (fd < 0) {
      pam_syslog(pamh, LOG_ERR, "unable to create %s", RENEWDCCS);
      free(newname);
      return;
    }
    fd = open(newname, O_CREAT|O_WRONLY, 0600);
  }
  if (fd < 0) {
    pam_syslog(pamh, LOG_ERR, "unable to create %s", newname);
    free(newname);
    return;
  }
  dprintf(fd, "%lu\n", (unsigned long)uid);
  free(newname);
  close(fd);
}

// for names like /run/user/%U/krbcc_%U 
// make need to create directory, but only if it's user-specific

void
assure_dir(char *template, char *filename, struct passwd *pwd) {
  char *cp;
  char *lastslash = NULL;
  char *nexttolastslash = NULL;
  int ok = 0;
  int err;

  // first, see if there's a % in the last directory compontent
  // if so, that's a user-specific directory, and we're willing to
  // create it. Otherwise, nothing we can do.

  for (cp = template; *cp; cp++) {
    if (*cp == '/') {
      nexttolastslash = lastslash;
      lastslash = cp;
    }
  }
  // is there a % between next to last and last?
  if (lastslash && nexttolastslash) {
    cp = strchr(nexttolastslash, '%');
    if (cp < lastslash)
      ok = 1;
  }
  
  // if not, nothing to do
  if (!ok)
    return;

  // there is. need to check the directory
  
  // get directory, by cutting off at least /
  cp = strrchr(filename, '/');
  *cp = '\0';
  if (mkdir(filename, 0700)) {
    // nothing to do
    *cp = '/';
    return;
  }
  // created it, set owner
  chown(filename, pwd->pw_uid, pwd->pw_gid);
  *cp = '/';

}

char *
build_cache_name(char *arg, uid_t uid, const char *username)
{
    char *cache_name = NULL;
    int retval;
    size_t len = 0, delta;
    char *p, *q;

    // compute length of final product
    for (p = arg; *p != '\0'; p++) {
      if (p[0] == '%' && p[1] == 'U') {
	len += snprintf(NULL, 0, "%ld", (long) uid);
	p++;
      } else if (p[0] == '%' && p[1] == 'u') {
	len += snprintf(NULL, 0, "%s", username);
	p++;
      } else {
	len++;
      }
    }
    len++;

    // now do it for real
    cache_name = malloc(len);
    if (cache_name == NULL) {
      return NULL;
    }
    for (p = arg, q = cache_name; *p != '\0'; p++) {
      if (p[0] == '%' && p[1] == 'U') {
	delta = snprintf(q, len, "%ld", (long) uid);
	q += delta;
	len -= delta;
	p++;
      } else if (p[0] == '%' && p[1] == 'u') {
	delta = snprintf(q, len, "%s", username);
	q += delta;
	len -= delta;
	p++;
      } else {
	*q = *p;
	q++;
	len--;
      }
    }
    *q = '\0';
    return cache_name;
}

PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv) {
  const char *ccname = pam_getenv(pamh, "KRB5CCNAME");
  int i;
  const char *username;
  char *fullname = NULL;
  char *cccopy = NULL;
  int ret;
  krb5_context context = NULL;
  int usecollection = 0;
  int fakename = 0;
  char *default_realm = NULL;
  const char *default_name = NULL;
  krb5_data realm_data;
  char *finalcred = NULL;
  char *tempcred = NULL;
  krb5_ccache cachecopy = NULL;
  krb5_ccache firstcache = NULL;
  krb5_cc_cursor cur = NULL;
  krb5_creds creds;
  struct passwd * pwd = NULL;
  struct passwd pwd_struct;
  char pwd_buf[2048];
  krb5_principal userprinc = NULL;
  int found_current_tgt = FALSE;
  time_t now;
  krb5_deltat minlife;
  char *minlife_st;
  const void *getcred;
  int iscron = 0;
  char *key;
  uid_t olduid;
  gid_t oldgid;
  const char *servicename;

  pam_syslog(pamh, LOG_INFO, "registering ccname %s", ccname);

  for (i = 0; i < argc; i++) {
    if (strcmp(argv[i], "usecollection") == 0)
      usecollection = 1;
    if (strcmp(argv[i], "usedefaultname") == 0)
      fakename = 1;
  }

  if (!fakename && !ccname) 
    // KRB5CCNAME not specified and we haven't been asked to fake it
    return PAM_SUCCESS;  // nothing to do

  // generally credentials will be set up by sssd, pam_krb5, or sshd
  // in all of those cases we want the usual code. However pam_kgetcred
  // will put credentials in /tmp, and we want them to stay there even
  // if the default is keyring. So iscron really means is_kgetcred
  if (pam_get_data(pamh, "kgetcred_test", &getcred) == PAM_SUCCESS)
    iscron = 1;

  // get basic user and kerberos info

  ret = krb5_init_context(&context);
  if (ret) goto err1;

  if (pam_get_user(pamh, &username, NULL) != PAM_SUCCESS) 
    return PAM_AUTHINFO_UNAVAIL;

  getpwnam_r(username, &pwd_struct, pwd_buf, sizeof(pwd_buf), &pwd);
  if (!pwd) goto err;

  olduid = getuid();
  oldgid = getgid();

  //
  // 1. If for some reason KRB5CCNAME isn't set, and config asks for it,
  // set KRB5CCNAME to the default ccache name. Renewd works better if
  // KRB5CCNAME is always set.
  //

  if (fakename && !ccname) {
    // KRB5CCNAME not specified, but we've been told to fake it.
    // Generate the user's default ccname.
    // Need to change to the user, since the Kerberos libraries get %{uid}
    // from the current uid.

    setresgid(pwd->pw_gid, pwd->pw_gid, -1);
    setresuid(pwd->pw_uid, pwd->pw_uid, -1);

    ccname =  krb5_cc_default_name(context);

    // now put back our real uid
    setresuid(olduid, olduid, -1);
    setresgid(oldgid, oldgid, -1);

    if (!ccname) {
      // no name, nothing to do
      krb5_free_context(context);
      pam_syslog(pamh, LOG_INFO, "krb5_cc_default_name returns null");
      return PAM_SUCCESS;  // nothing to do      
    }

    pam_syslog(pamh, LOG_INFO, "registering default ccname for this user %s", ccname);    

  }

  //
  // 2. If ccache is /tmp and default is KEYRING, move the cache into the KEYRING.
  // sshd in some versions ignores krb5.conf and puts the cache into /tmp. First, we
  // want a consistent user experience. Second, gssd seems to work better with the
  // keyring. ccache names for keyring always have to start KEYRING, so we don't need
  // fancy normalization code for the check.
  //

  setresgid(pwd->pw_gid, pwd->pw_gid, -1);
  setresuid(pwd->pw_uid, pwd->pw_uid, -1);
  default_name =  krb5_cc_default_name(context);  

  setresuid(olduid, olduid, -1);
  setresgid(oldgid, oldgid, -1);

  ret = krb5_get_default_realm(context, &default_realm);
  if (ret) goto err;
  ret = krb5_build_principal(context, &userprinc, strlen(default_realm), default_realm, username, NULL);
  if (ret) goto err1;

  // For NFS to work, rpc.gssd has to be able to find the credentials. It looks at /tmp and
  // the default in krb5.conf. We've got two issues: (1) sshd likes to put things in /tmp
  // (2) Ubuntu sssd likes to put things in keyring, even if that's not right. If krb5.conf
  // points to KCM and the actual ticket is in KEYRING, gssd will fail and the user won't be
  // able to get files. So if the default is not /tmp, and login put the credential somewhere
  // other than the default, move it.
  
  if (!iscron && default_name &&
      is_collection_type(default_name) &&
      strcmp(get_cc_type(ccname), get_cc_type(default_name)) != 0) {
    const char *cp;
    int numcolon = 0; 
    char *prop = NULL;


    setresgid(pwd->pw_gid, pwd->pw_gid, -1);
    setresuid(pwd->pw_uid, pwd->pw_uid, -1);

    ret = krb5_cc_resolve(context, ccname, &firstcache);
    if (ret) goto err2;

    // If possible we reuse an existing cache, if one has the right principal.
    // There is no API call that lets us look only for caches of a specific type. The best we
    // can do is this. We then have to verify that it's the right type. If it's not
    // we ignore it. That means in some cases we can create a new cache every time
    // we login. That's not so terrible. sshd and sssd already do that.
    ret = krb5_cc_cache_match(context, userprinc, &cachecopy);
    if (ret == 0 && strcmp(krb5_cc_get_type(context, cachecopy), get_cc_type(default_name)) != 0)
      ret = 1;

    // if none, create one
    if (ret) {
      char * newtype = get_cc_type(default_name);

      ret = krb5_cc_new_unique(context, newtype, NULL, &cachecopy);
      if (ret)   pam_syslog(pamh, LOG_INFO, "new_uniq failed %s\n", newtype);


      // it's not clear whether this is a good idea, but it's probably more
      //   likely to be right than wrong. If there is an existing cache and
      //   it's not primary, we leave things to avoid conflicting with something
      //   the user has done explicitly. But if there isn't one and we have to
      //   create it, it's probably best to make it primary
      // make it primary. ignore failure
      if (!ret)
	krb5_cc_switch(context, cachecopy); 

    }

    if (ret) goto err2;
    ret = krb5_cc_initialize(context, cachecopy, userprinc);
    if (ret) goto err2;
    ret = krb5_cc_copy_creds(context, firstcache, cachecopy);
    if (ret) goto err2;

    ret = krb5_cc_get_full_name(context, cachecopy, &tempcred);
    if (ret) goto err2;

    ccname = tempcred;  // tempcred will be freed at exit

    krb5_cc_close(context, cachecopy);
    cachecopy = NULL;
    krb5_cc_destroy(context, firstcache);
    firstcache = NULL;

    // now put back our real uid
    setresuid(olduid, olduid, -1);
    setresgid(oldgid, oldgid, -1);

    // reset environment to collection
    if (asprintf(&prop, "%s=%s", "KRB5CCNAME", ccname) > 0) {
	pam_putenv(pamh, prop);
	if (prop)
	  free(prop);
    }

    pam_syslog(pamh, LOG_INFO, "moving to %s", tempcred);

  }    

  //
  // 3. Normalize cache name. For /tmp, there's no issue as there are no
  // collections. For KEYRING, we do two things (1) if KRB5CCNAME was set
  // to the collection, find the actual ccache and put it in ccname. Leave
  // KRB5CCNAME to the collection, but we need the actual ccache for some
  // later codr. (2) if KRB5CCNAME is set to an actual ccache, reset it
  // to the collection. We want users to get a consistent experience, and to
  // be able to use kswitch.
  //


  // for sss, we need to find the cache for the current principal. it's not necesarily the primary
  //   if another process did kswitch
  // for ssh, change KRB5CCNAME to the collection, or kinit with another principal will clobber this one

  if (is_collection(ccname)) {
      krb5_ccache ccache = NULL;

      setresgid(pwd->pw_gid, pwd->pw_gid, -1);
      setresuid(pwd->pw_uid, pwd->pw_uid, -1);

      // find cache that matches principal
      ret = krb5_cc_cache_match(context, userprinc, &ccache);
      if (ret) goto err1;      

      ret = krb5_cc_get_full_name(context, ccache, &fullname);
      if (ret) goto err1;      
      
      // make it primary
      // ret = krb5_cc_switch(context, ccache); 
      // if (ret) goto err1;

    err1:
      if (ret)
	pam_syslog(pamh, LOG_ERR, "%s", error_message(ret));

      if (ccache)
	krb5_cc_close(context, ccache);

      setresuid(olduid, olduid, -1);
      setresgid(oldgid, oldgid, -1);

      // end of collection
  } else if (is_collection_type(ccname) && usecollection) {
      // have specific cache
      char *prop = NULL;
      char *collection = convert_to_collection(ccname, (uid_t)-1);

      // reset environment to collection
      if (asprintf(&prop, "KRB5CCNAME=%s", collection) > 0) {
	// ccname will no longer be valid after the putenv
	cccopy = malloc(strlen(ccname) + 1);
	strcpy(cccopy, ccname);

	pam_putenv(pamh, prop);
      }
      if (prop)
	free(prop);
      free(collection);
  }

  // note: ccname doesn't have to be freed; fullname does
  if (fullname) {
    ccname = fullname;
  } else if (cccopy) {
    ccname = cccopy;
  }

  //
  // 4. register the cache for renewal
  //

  register_for_delete(pamh, ccname, pwd->pw_uid);

  //
  // 5. except for cron, see if the ticket lifetime is too small, and warn user.
  // Doesn't make sense for cron because there's no user terminal for warning.
  // cron sets PAM_SILENT, which seems like the best test to use
  //

  if (!(flags & PAM_SILENT)) {

  realm_data.data = default_realm;
  realm_data.length = strlen(default_realm);
  
  // see if it's got enough time left
  krb5_appdefault_string(context, "register-cc", &realm_data, "ticket_minlife", "0m", &minlife_st);
  ret = krb5_string_to_deltat(minlife_st, &minlife);
  if (ret)
    goto err;

  if (minlife) {

    // KCM requires cc_resolve to be done as user
    setresgid(pwd->pw_gid, pwd->pw_gid, -1);
    setresuid(pwd->pw_uid, pwd->pw_uid, -1);
    ret = krb5_cc_resolve(context, ccname, &firstcache);
    setresuid(olduid, olduid, -1);
    setresgid(oldgid, oldgid, -1);

    if (ret) goto err;
    if ((ret = krb5_cc_start_seq_get(context, firstcache, &cur))) {
      goto err;
    }
    found_current_tgt = FALSE;   // found a TGT that is current
    
    now = time(0);
    
    while (!(ret = krb5_cc_next_cred(context, firstcache, &cur, &creds))) {
      // only renewable creds are worth looking at. and must be TGT
      if ((creds.ticket_flags & TKT_FLG_RENEWABLE) && is_local_tgt(creds.server, &userprinc->realm)) {
	// enough time left, it's current
	if (((time_t)(uint32_t)creds.times.endtime - now) > minlife) {
	  found_current_tgt = TRUE;
	  krb5_free_cred_contents(context, &creds);
	  break;
	}
	krb5_free_cred_contents(context, &creds);
      }
    }
    krb5_cc_end_seq_get(context, firstcache, &cur);
    
    // if ticket isn't good enough issue warning.
    // most of the code is for reading warning from a file configured in krb5.conf
    // the text is probably a bit long to put in krb5.conf directly
    if (!found_current_tgt) {
      char *warnfilename = NULL;
      FILE *warnfile = NULL;
      char *warntext = NULL;
      long fsize;
      int warnfd;

      krb5_appdefault_string(context, "register-cc", &realm_data, "ticket_warn_file", "", &warnfilename);      
      // do we have a file with text? Otherwise there's a builtin default text
      if (warnfilename && strlen(warnfilename) > 0) {
	warnfile = fopen(warnfilename, "r");
	if (!warnfile)
	  goto texterr;
	// you'd expect to use fstat, but it's missing from libc somehow
	fseek(warnfile, 0, SEEK_END);
	fsize = ftell(warnfile);
	fseek(warnfile, 0, SEEK_SET);  //same as rewind(f);
	warntext = malloc(fsize + 1);
	if (!warntext)
	  goto texterr;
	if (fread(warntext, fsize, 1, warnfile) != 1)
	  goto texterr;
	warntext[fsize] = 0;
	goto textok;
	
      texterr:
	if (warntext) {
	  free(warntext);
	  warntext = NULL;
	}
      textok:
	if (warnfile) 
	  fclose(warnfile);
      }

      if (warntext) {
	pam_info(pamh, "%s", warntext);
	free(warntext);
      } else
	pam_info(pamh, "\n**********************************************************************\nYour Kerberos ticket doesn't have enough lifetime left. You may lose\naccess to your files during this session. We suggest using the\ncommand \"kinit\" to get a new ticket.\n**********************************************************************\n");
    }
  }

  } // end if iscron

 err:
  if (ret)
    pam_syslog(pamh, LOG_ERR, "%s", error_message(ret));
  if (ret)
    printf("%s\n", error_message(ret));
  if (userprinc)
    krb5_free_principal(context, userprinc);
  if (cachecopy)
    krb5_cc_close(context, cachecopy);
  if (tempcred)
    free(tempcred);
  if (finalcred)
    free(finalcred);
  if (firstcache)
    krb5_cc_close(context,firstcache);
  if (default_realm)
    krb5_free_default_realm(context, default_realm);
  if (cccopy)
    free(cccopy);
  if (fullname)
    krb5_free_string(context, fullname);
  if (context)
    krb5_free_context(context);

  return PAM_SUCCESS;

 err2:
  setresuid(olduid, olduid, -1);
  setresgid(oldgid, oldgid, -1);
  goto err;

}

PAM_EXTERN int pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char **argv) {
  return PAM_SUCCESS;
} 

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv) {
  return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv) {
  pam_sm_open_session(pamh, flags, argc, argv);
  return PAM_SUCCESS; // optional so always ok
}

